processor/app.go

```diff
@@ -0,0 +1,125 @@
+package processor
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/ONSdigital/census-rm-pubsub-adapter/models"
+	"log"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/pkg/errors"
+	"github.com/streadway/amqp"
+)
+
+type App struct {
+	RabbitConn   *amqp.Connection
+	RabbitChan   *amqp.Channel
+	PubSubClient *pubsub.Client
+	EqReceiptSub *pubsub.Subscription
+	MessageChan  chan pubsub.Message
+}
+
+func New(ctx context.Context, rabbitConnectionString, projectID string) *App {
+
+	//set up rabbit connection
+	var err error
+	a := &App{}
+	a.RabbitConn, err = amqp.Dial(rabbitConnectionString)
+	failOnError(err, "Failed to connect to RabbitMQ")
+
+	a.RabbitChan, err = a.RabbitConn.Channel()
+	failOnError(err, "Failed to open a channel")
+
+	//setup pubsub connection
+	a.PubSubClient, err = pubsub.NewClient(ctx, projectID)
+	failOnError(err, "Pubsub client creation failed")
+
+	//setup subcriptions
+	a.EqReceiptSub = a.PubSubClient.Subscription("rm-receipt-subscription")
+	a.MessageChan = make(chan pubsub.Message)
+
+	return a
+}
+
+func (a *App) Process(ctx context.Context) {
+	for {
+		select {
+		case msg := <-a.MessageChan:
+			var eqReceiptReceived models.EqReceipt
+			err := json.Unmarshal(msg.Data, &eqReceiptReceived)
+			if err != nil {
+				// TODO Log the error and DLQ the message when unmarshalling fails, printing it out is a temporary solution
+				log.Printf("Error unmarshalling message: %q, error: %q\n", string(msg.Data), err)
+				msg.Ack()
+				return
+			}
+
+			log.Printf("Got QID: %q\n", eqReceiptReceived.Metadata.QuestionnaireID)
+			rmMessageToSend, err := convertEqReceiptToRmMessage(&eqReceiptReceived)
+			if err != nil {
+				log.Println(errors.Wrap(err, "failed to convert receipt to message"))
+			}
+			sendRabbitMessage(rmMessageToSend, a.RabbitChan)
+			msg.Ack()
+		}
+	}
+
+}
+
+func (a *App) Consume(ctx context.Context) {
+	log.Println("Launched PubSub message listener")
+
+	err := a.EqReceiptSub.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
+		log.Printf("Got message: %q\n", string(msg.Data))
+		a.MessageChan <- *msg
```

Comment:
so I presume this operation blocks if the channel buffer is full?

Reply:
that is correct, eventually we would have multiple pubsub go routines writing to that chan so maybe we can set the chan buffer size to the amount of routines that have been configured....but that will come at a later point when we enable that kind of configuration